File: 0716-max-stack/0716-max-stack.cpp
// SC- O(n)
class MaxStack {
private:
    priority_queue<pair<int,int>> max_heap;
    stack<pair<int, int>> st;
    int cnt; // to assign unique id to each new elem
    unordered_set<int> removed;

public:
    MaxStack() {
        cnt = 0;
    }
    
    void push(int x) {
        // TC- O(logx)
        max_heap.emplace(x, cnt);
        st.emplace(x, cnt);
        cnt++;
    }
    
    int pop() {
        // TC- amortized O(1) since one elem can be removed only once from stack

        // lazy update
        while(removed.count(st.top().second))
            st.pop();
        
        // add to removal list and remove from stack
        pair<int,int> toremove = st.top();
        st.pop();
        removed.insert(toremove.second);
        return toremove.first;
    }
    
    int top() {
        // TC- O(1)
        // lazy update
        while(removed.count(st.top().second))
            st.pop();

        return st.top().first;
    }
    
    int peekMax() {
        // TC- O(1)
        // lazy update
        while (removed.count(max_heap.top().second))
            max_heap.pop();
        return max_heap.top().first;
    }
    
    int popMax() {
        // TC- amortized O(1) since one elem can be removed only once from stack

        // lazy update
        while (removed.count(max_heap.top().second))
            max_heap.pop();

        pair<int, int> toremove = max_heap.top();
        max_heap.pop();
        removed.insert(toremove.second);
        return toremove.first;
    }
};

/**
 * Your MaxStack object will be instantiated and called as such:
 * MaxStack* obj = new MaxStack();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->peekMax();
 * int param_5 = obj->popMax();
 */

File: 0716-max-stack/0716-max-stack.py
class MaxStack:

    def __init__(self):
        # stack to peek or pop last added element quickly
        # priority queue to peep or pop max element quickly
        # but element removed from stack should be updated in pq. but to do that you need to iterate through it. so just do lazy deletion
        # O(n)
        self.stack = []
        self.heap = []
        self.to_remove = set()
        self.cnt = 0 # to assign new id to each new element

    def push(self, x: int) -> None:
        # O(logn) = O(1)+O(logn)
        self.stack.append((x, self.cnt))
        heapq.heappush(self.heap, (-x, -self.cnt)) # since default is minheap
        self.cnt += 1

    def pop(self) -> int:
        # O(logn)- amortized
        while self.stack and self.stack[-1][1] in self.to_remove:
            self.stack.pop()
        num, idx = self.stack.pop()
        self.to_remove.add(idx)
        return num

    def top(self) -> int:
        # O(1)- amortized
        while self.stack and self.stack[-1][1] in self.to_remove:
            self.stack.pop()
        return self.stack[-1][0]

    def peekMax(self) -> int:
        while self.heap and -self.heap[0][1] in self.to_remove:
            heapq.heappop(self.heap)
        return -self.heap[0][0]

    def popMax(self) -> int:
        while self.heap and -self.heap[0][1] in self.to_remove:
            heapq.heappop(self.heap)
        num, idx = heapq.heappop(self.heap)
        self.to_remove.add(-idx)
        return -num

# Your MaxStack object will be instantiated and called as such:
# obj = MaxStack()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.peekMax()
# param_5 = obj.popMax()